File: packages/nitro-renderer/src/assets/GraphicAssetCollection.ts
import type {
    IAsset,
    IAssetData,
    IAssetPalette,
    IGraphicAsset,
    IGraphicAssetCollection,
    IGraphicAssetPalette,
} from '@nitrodevco/nitro-api';
import type { Texture, TextureSource } from 'pixi.js';

import { GraphicAsset } from './GraphicAsset';
import { GraphicAssetPalette } from './GraphicAssetPalette';

export class GraphicAssetCollection implements IGraphicAssetCollection {
    private static PALETTE_ASSET_DISPOSE_THRESHOLD: number = 10;

    private _referenceCount: number = 0;

    private _type: string;
    private _textureSource: TextureSource | undefined;
    private _data: IAssetData;
    private _textures: Map<string, Texture>;
    private _assets: Map<string, IGraphicAsset>;
    private _palettes: Map<number, IGraphicAssetPalette>;
    private _paletteAssetNames: string[];

    constructor(data: IAssetData, source: TextureSource | undefined, textures: Record<string, Texture> | undefined) {
        if (!data) throw new Error('invalid_collection');

        this._type = data.type;
        this._textureSource = source;
        this._data = data;
        this._textures = new Map();
        this._assets = new Map();
        this._palettes = new Map();
        this._paletteAssetNames = [];

        this.addLibraryAsset(textures);
        this.define(data);
    }

    public static removeFileExtension(name: string): string {
        return name.substring(0, name.lastIndexOf('.')) || name;
    }

    public dispose(): void {
        if (this._palettes) this._palettes.clear();

        if (this._paletteAssetNames) {
            this.disposePaletteAssets();

            this._paletteAssetNames = [];
        }

        if (this._assets) {
            for (const asset of this._assets.values()) asset.recycle();

            this._assets.clear();
        }
    }

    public define(data: IAssetData): void {
        const assets = data.assets;
        const palettes = data.palettes;

        if (assets) this.defineAssets(assets);

        if (palettes) this.definePalettes(palettes);
    }

    public getAssetWithPalette(name: string, paletteId: number): IGraphicAsset | undefined {
        const saveName = name + '@' + paletteId;

        let asset = this.getAsset(saveName);

        if (asset) return asset;

        asset = this.getAsset(name);

        if (!asset || !asset.texture || !asset.usesPalette) return undefined;

        const palette = this.getPalette(paletteId);

        if (!palette) return asset;

        const texture = palette.applyPalette(asset.texture);

        if (!texture) return asset;

        this._paletteAssetNames.push(saveName);

        return this.addAsset(
            saveName,
            texture,
            asset.x,
            asset.y,
            asset.flipH,
            asset.flipV,);
    }

    public getPaletteColors(paletteId: number): number[] {
        const palette = this.getPalette(paletteId);

        if (palette) return [palette.primaryColor, palette.secondaryColor];

        return [0, 0];
    }

    public addAsset(
        name: string,
        texture: Texture,
        x: number = 0,
        y: number = 0,
        flipH: boolean = false,
        flipV: boolean = false,
        usesPalette: boolean = false,
        replace: boolean = false
    ): IGraphicAsset | undefined {
        if (!name || !texture) return undefined;

        let asset = this._assets.get(name) as GraphicAsset;

        if (asset) {
            if (!replace) return undefined;

            this.disposeAsset(name);
        }

        asset = GraphicAsset.createAsset(name, name, texture, x, y, flipH, flipV, usesPalette);

        this._assets.set(name, asset);

        return asset;
    }

    public disposeAsset(name: string): void {
        const asset = this._assets.get(name);

        if (!asset || !this._assets.delete(name)) return;

        if (asset.texture) {

            const isSubTexture = asset.texture.frame.x !== 0 ||
                asset.texture.frame.y !== 0 ||
                asset.texture.frame.width !== asset.texture.source.width ||
                asset.texture.frame.height !== asset.texture.source.height;

            if (!isSubTexture) asset.texture.destroy(true);
        }

        asset.recycle();
    }

    public getAsset(name: string): IGraphicAsset | undefined {
        return this._assets.get(name);
    }

    public getTexture(name: string): Texture | undefined {
        return this._textures.get(name);
    }

    public getPalette(paletteId: number): IGraphicAssetPalette | undefined {
        return this._palettes.get(paletteId);
    }

    public addReference(): void {
        this._referenceCount++;
    }

    public removeReference(): void {
        this._referenceCount--;

        if (this._referenceCount <= 0) {
            this._referenceCount = 0;

            this.disposePaletteAssets(false);
        }
    }

    public get referenceCount(): number {
        return this._referenceCount;
    }

    public get name(): string {
        return this._type;
    }

    public get data(): IAssetData {
        return this._data;
    }

    public get textureSource(): TextureSource | undefined {
        return this._textureSource;
    }

    public get textures(): Map<string, Texture> {
        return this._textures;
    }

    public get assets(): Map<string, IGraphicAsset> {
        return this._assets;
    }

    public get palettes(): Map<number, IGraphicAssetPalette> {
        return this._palettes;
    }

    private defineAssets(assets: IAsset[]): void {
        for (const asset of assets) {
            const source = asset.source ?? asset.name;
            const x = -(asset.x || 0);
            const y = -(asset.y || 0);
            const flipH = asset.flipH ?? false;
            const flipV = asset.flipV ?? false;
            const usesPalette = asset.usesPalette ?? false;
            const texture = this.getTexture(source);

            if (!texture) continue;

            this._assets.set(asset.name, GraphicAsset.createAsset(asset.name, source, texture, x, y, flipH, flipV, usesPalette));
        }
    }

    private definePalettes(palettes: IAssetPalette[]): void {
        for (const palette of palettes) {
            if (this._palettes.get(palette.id)) continue;

            let colorOne = 0xffffff;
            let colorTwo = 0xffffff;

            let color = palette.color1;

            if (color && color.length > 0) colorOne = parseInt(color, 16);

            color = palette.color2;

            if (color && color.length > 0) colorTwo = parseInt(color, 16);

            if (palette.rgb) this._palettes.set(palette.id, new GraphicAssetPalette(palette.rgb, colorOne, colorTwo));
        }
    }

    private addLibraryAsset(textures: Record<string, Texture> | undefined): void {
        if (!textures) return;

        for (const name in textures) {
            const texture = textures[name];

            if (!texture) continue;

            this._textures.set(GraphicAssetCollection.removeFileExtension(name), texture);
        }
    }

    private disposePaletteAssets(disposeAll: boolean = true): void {
        if (this._paletteAssetNames) {
            if (disposeAll || this._paletteAssetNames.length > GraphicAssetCollection.PALETTE_ASSET_DISPOSE_THRESHOLD) {
                for (const name of this._paletteAssetNames) this.disposeAsset(name);

                this._paletteAssetNames = [];
            }
        }
    }
}

File: packages/nitro-swift/Sources/NitroAssets/GraphicAssetCollection.swift
import Foundation
import SpriteKit

/// Swift port of `GraphicAssetCollection` (packages/nitro-renderer/src/assets/GraphicAssetCollection.ts).
/// One instance per loaded `.nitro` bundle (a furniture item, an avatar part library, etc.).
public final class GraphicAssetCollection {
    private static let paletteAssetDisposeThreshold = 10

    public private(set) var referenceCount: Int = 0

    /// `data.type` - also the key this collection is registered under in `AssetManager.collections`.
    public let name: String
    public let data: AssetDataCore
    /// The full merged bundle JSON (all non-spritesheet manifest files, shallow-merged), kept so
    /// NitroAvatar/NitroRoom can decode the richer fields (`logic`, `visualizations`,
    /// `roomVisualization`, `animations`, ...) this layer doesn't need to know about.
    public let rawJSON: [String: Any]
    public let textureSheet: SKTexture?

    private var _textures: [String: SKTexture] = [:]
    private var _assets: [String: GraphicAsset] = [:]
    private var _palettes: [Int: GraphicAssetPalette] = [:]
    private var paletteAssetNames: [String] = []

    public init(data: AssetDataCore, rawJSON: [String: Any], libraryTextures: [String: SKTexture], textureSheet: SKTexture? = nil) {
        self.data = data
        self.name = data.type
        self.rawJSON = rawJSON
        self.textureSheet = textureSheet

        addLibraryAssets(libraryTextures)
        define(data)
    }

    public static func removeFileExtension(_ name: String) -> String {
        guard let idx = name.lastIndex(of: ".") else { return name }

        return String(name[name.startIndex..<idx])
    }

    public func dispose() {
        _palettes.removeAll()
        disposePaletteAssets(disposeAll: true)
        paletteAssetNames.removeAll()

        for asset in _assets.values { asset.recycle() }

        _assets.removeAll()
    }

    public func define(_ data: AssetDataCore) {
        if let assets = data.assets { defineAssets(assets) }
        if let palettes = data.palettes { definePalettes(palettes) }
    }

    public func getAssetWithPalette(_ name: String, paletteId: Int) -> GraphicAsset? {
        let saveName = "\(name)@\(paletteId)"

        if let existing = getAsset(saveName) { return existing }

        guard let asset = getAsset(name), let texture = asset.texture, asset.usesPalette else { return nil }
        guard let palette = getPalette(paletteId) else { return asset }

        let recolored = palette.applyPalette(texture)

        paletteAssetNames.append(saveName)

        return addAsset(name: saveName, texture: recolored, x: asset.x, y: asset.y, flipH: asset.flipH, flipV: asset.flipV)
    }

    public func getPaletteColors(_ paletteId: Int) -> (primary: UInt32, secondary: UInt32) {
        guard let palette = getPalette(paletteId) else { return (0, 0) }

        return (palette.primaryColor, palette.secondaryColor)
    }

    @discardableResult
    public func addAsset(
        name: String,
        texture: SKTexture,
        x: Double = 0,
        y: Double = 0,
        flipH: Bool = false,
        flipV: Bool = false,
        usesPalette: Bool = false,
        replace: Bool = false
    ) -> GraphicAsset? {
        if _assets[name] != nil {
            if !replace { return nil }

            disposeAsset(name)
        }

        let asset = GraphicAsset(name: name, source: name, texture: texture, x: x, y: y, flipH: flipH, flipV: flipV, usesPalette: usesPalette)

        _assets[name] = asset

        return asset
    }

    public func disposeAsset(_ name: String) {
        guard let asset = _assets.removeValue(forKey: name) else { return }

        asset.recycle()
    }

    public func getAsset(_ name: String) -> GraphicAsset? { _assets[name] }
    public func getTexture(_ name: String) -> SKTexture? { _textures[name] }
    public func getPalette(_ paletteId: Int) -> GraphicAssetPalette? { _palettes[paletteId] }

    public func addReference() { referenceCount += 1 }

    public func removeReference() {
        referenceCount -= 1

        if referenceCount <= 0 {
            referenceCount = 0

            disposePaletteAssets(disposeAll: false)
        }
    }

    public var textures: [String: SKTexture] { _textures }
    public var assets: [String: GraphicAsset] { _assets }
    public var palettes: [Int: GraphicAssetPalette] { _palettes }

    private func defineAssets(_ list: [AssetDefinition]) {
        for asset in list {
            let source = asset.source ?? asset.name
            let x = -(asset.x ?? 0)
            let y = -(asset.y ?? 0)

            guard let texture = getTexture(source) else { continue }

            _assets[asset.name] = GraphicAsset(
                name: asset.name, source: source, texture: texture, x: x, y: y,
                flipH: asset.flipH ?? false, flipV: asset.flipV ?? false, usesPalette: asset.usesPalette ?? false
            )
        }
    }

    private func definePalettes(_ list: [AssetPaletteDefinition]) {
        for palette in list {
            if _palettes[palette.id] != nil { continue }

            var colorOne: UInt32 = 0xFFFFFF
            var colorTwo: UInt32 = 0xFFFFFF

            if let color = palette.color1, !color.isEmpty, let value = UInt32(color, radix: 16) { colorOne = value }
            if let color = palette.color2, !color.isEmpty, let value = UInt32(color, radix: 16) { colorTwo = value }

            if let rgb = palette.rgb { _palettes[palette.id] = GraphicAssetPalette(rgb: rgb, primaryColor: colorOne, secondaryColor: colorTwo) }
        }
    }

    private func addLibraryAssets(_ textures: [String: SKTexture]) {
        for (name, texture) in textures {
            _textures[GraphicAssetCollection.removeFileExtension(name)] = texture
        }
    }

    private func disposePaletteAssets(disposeAll: Bool = true) {
        guard disposeAll || paletteAssetNames.count > GraphicAssetCollection.paletteAssetDisposeThreshold else { return }

        for name in paletteAssetNames { disposeAsset(name) }

        paletteAssetNames.removeAll()
    }
}
